orttraining/tools/scripts/performance_investigation.py

```diff
@@ -0,0 +1,133 @@
+
+        # Look for node surrounded (sandwiched) by Casts node, see if ORT has already implemented fp16-saft kernels for them
+        if optimized_file:
+            if node.op_type == "Cast" and len(path) > 1 and path[-2].op_type == "Cast":
+                msgs.append(f"Excessive casts around {path[-1].name} node.")        
+
+        # Look for (Simplified)LayerNormalization in *_pre_grad_optimized_<mode>.onnx graph.
+        # The layernorm subgraph (search for Pow node to begin with) should be fused into a single node.
+        if pre_grad_optimized_file:
+            if node.op_type == "LayerNormalization" or node.op_type == "SimplifiedLayerNormalization":
+                for prev in path[::-1]:
+                    if prev.op_type == "Pow":
+                        msgs.append(f"Standalone {node.op_type} {node.name} found. The layernorm subgraph starting with {prev.name} should be fused into a single node.")
+                    break
+
+        # Look for (Fast)Gelu in *_pre_grad_optimized_<mode>.onnx graph.
+        # The gelu subgraph (search for Erf node to begin with) should be fused into a single node.
+        if pre_grad_optimized_file:
+            if node.op_type == "Gelu" or node.op_type == "FastGelu":
+                for prev in path[::-1]:
+                    if prev.op_type == "Erf":
+                        msgs.append(f"Standalone {node.op_type} {node.name} found. The gelu subgraph starting with {prev.name} should be fused into a single node.")
+                    break
+
+        # Look for stand-alone Dropout node in *_execution_model_<mode>.onnx graph.
+        # Examine whether it should be fused with surrounding Add ops into BiasDropout node.
+        if execution_model_file:
+            if node.op_type == "Dropout" and path and path[-1].op_type == "Add":
+                msgs.append(f"Examine whether {node.name} should be fused with the leading {path[-1].name} op into BiasDropout node.")
+
+        # Look for stand-alone Softmax node in *_execution_model_<mode>.onnx graph.
+        # Examine whether it should be fused with the leading Add ops into BiasSoftmax node.
+        if execution_model_file:
+            if node.op_type == "Softmax" and path and path[-1].op_type == "Add":
+                msgs.append(f"Examine whether {node.name} should be fused with the leading {path[-1].name} op into BiasSoftmax node.")
+
+        for o in node.output:
+            for next_node in input_to_nodes.get(o, []):
+                dfs(next_node, path + [node])
+
+    for i in model.graph.input:
+        if i.name in output_to_node:
+            continue
+        for node in input_to_nodes.get(i.name, []):
+            if node.name not in visited:
+                dfs(node, [])
+
+    for i in model.graph.input:
+        if i.name in output_to_node:
+            continue
+        for node in input_to_nodes.get(i.name, []):
+            if node.name not in visited:
+                dfs(node, [])
```

Comment:
As there are many connection edges between forwards nodes and backward nodes in a training graph, with dfs traversal, it's likely that you jump directly from forward node to a backward node, and miss the pattern you are looking for. 

Actually, there is no single graph transversal algo that give you a reliable serialization that you can match with a linear pattern.